source/unified-test-format/unified-test-format.rst

```diff
@@ -2492,7 +2592,223 @@ An example of this operation follows::
+  scheduled.
+
+- ``operation``: Required `operation`_ object. The operation to schedule on the
+  thread. This object must be a valid operation as described in `Entity Test
+  Operations`_.
+
+An example of this operation follows::
+
+     - name: runOnThread
+       object: testRunner
+       arguments:
+         thread: *thread0
+         operation:
+           name: insertOne
+           object: *collection0
+           arguments:
+             document: { _id: 2 }
+           expectResult:
+             $$unsetOrMatches:
+               insertedId: { $$unsetOrMatches: 2 }
+
+
+waitForThread
+~~~~~~~~~~~~~
+
+The "waitForThread" operation instructs the test runner to notify the given
+thread that no more operations are forthcoming, wait for it to complete its last
+operation, and assert that it exited without any errors.
 
+The following arguments are supported:
+
+- ``thread``: Required string. Thread entity that should be stopped and waited
+  for.
+
+An example of this operation follows::
+
+    - name: waitForThread
+      object: testRunner
+      arguments:
+        thread: *thread0
+  
+
+waitForEvent
+~~~~~~~~~~~~
+
+The "waitForEvent" operation instructs the test runner to wait until the
+specified MongoClient has published a specific event a given number of
+times. Note that this includes any events published before the waitForEvent
+operation started.
+
+The following arguments are suported:
+
+- ``client``: Required string. Client entity whose events the runner will be
+  waiting for.
+
+- ``event``: Required `expectedEvent`_ object. The event which the test runner
+  is waiting to be observed on the provided client. The assertions laid out in
+  `expectedEvent`_ MUST be used to determine if an observed event matches
+  ``event``. ``event`` MAY have an event type that was not included in the
+  ``client``'s ``observeEvents`` field.
+
+- ``count``: Required integer. The number of matching events to wait for before
+  resuming execution of subsequent operations.
+
+For example, the following instructs the test runner to wait for at least one
+poolClearedEvent to be published::
+
+    - name: waitForEvent
+      object: testRunner
+      arguments:
+        client: *client0
+        event:
+          poolClearedEvent: {}
+        count: 1
+
+
+assertEventCount
+~~~~~~~~~~~~~~~~
+
+The "assertEventCount" operation instructs the test runner to assert the
+specified MongoClient has published a specific event a given number of times so
+far in the test.
+
+The following arguments are supported:
+
+- ``client``: Required string. Client entity whose events the runner will be
+  counting.
+
+- ``event``: Required `expectedEvent`_ object. The event which the test runner
+  will be counting on the provided client. The assertions laid out in
+  `expectedEvent`_ MUST be used to determine if an observed event matches
+  ``event``. ``event`` MAY have an event type that was not included in the
+  ``client``'s ``observeEvents`` field.
```

Comment:
Given that explanation, I think this could be problematic for a number of test runners. I can tell you that PHPLIB only collects events that are explicitly observed (i.e. `observeEvents` in the `client` entity definition). Observed events are appended to a list, which is then accessed later when processing `test.expectEvents`. This list is the same as what is used for special test operations such as [`assertDifferentLsidOnLastTwoCommands`](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#assertdifferentlsidonlasttwocommands), which is why the spec notes that "The client entity SHOULD include "commandStartedEvent" in observeEvents." when using the operation.

> The existing SDAM integration tests only assert on command events and only ever wait on SDAM/CMAP events

For this particular case, the existing test format should suffice. [`test.expectEvents`](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#test-expectevents) is an array, such that events of different types would be listed separately. In fact, there is no way (through `test.expectEvents` at least) to assert on the order of events of different types (e.g. a particular SDAM event after another APM event).

If you need to accomplish what is done in [`pool-checkout-returned-connection-maxConnecting.yml`](https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/tests/pool-checkout-returned-connection-maxConnecting.yml#L43) (i.e. observing all CMAP events for purposes of `waitForEvent` but ignoring some for `expectEvents`), I think the best course of action would be to extend [`expectedEventsForClient`](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#expectedeventsforclient) like we did with `ignoreExtraEvents`. The new field could be called `ignoreEvents` and contain a list of event names that should be skipped during the assertion against the expected list.

To take a step back, I'm arguing that `observeEvents` in the `client` definition is not only intended for `test.expectEvents` but is more generally about configuring event subscribers for _anything_ the test might need to do. If we could reach some solution that honors that, I think it'd avoid any surprises for test runner implementations (I'm more worried about others than PHPLIB, since I've already wrapped my head around this).

Reply:
> To take a step back, I'm arguing that observeEvents in the client definition is not only intended for test.expectEvents but is more generally about configuring event subscribers for anything the test might need to do

This interpretation makes sense to me. Rust records all events and uses "observeEvents" more as a post-processing filter, but I can see why all drivers don't do it that way. Also, adding `ignoreEvents` to the actual assertions seems like a more direct solution anyways. Updated to say the event type SHOULD be one included in the client's observeEvents field.

Reply:
Do we still need to add `ignoreEvents` to [`expectedEventsForClient`](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#expectedeventsforclient), or was that only hypothetical if CMAP tests were to be ported to the unified test format? I assume the absence of a ticket in [DRIVERS-1525](https://jira.mongodb.org/browse/DRIVERS-1525) means there are no plans to do so.

If nothing else is required, feel free to resolve this thread.

Reply:
The SDAM integration tests don't require it. The original idea was more of a forward thinking thing, so it's not needed for now.